pkg/types/error.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+)
+
+type ErrorCode string
+
+const (
+	ErrorCodeResultUnknown                   = ErrorCode("ResultUnknown")
+	ErrorCodeFunctionFailedWithoutRollback   = ErrorCode("FunctionFailedWithoutRollback")
+	ErrorCodeFunctionFailedRollbackSucceeded = ErrorCode("FunctionFailedRollbackSucceeded")
+	ErrorCodeFunctionFailedRollbackFailed    = ErrorCode("FunctionFailedRollbackFailed")
+)
+
+type Error struct {
+	Code    ErrorCode `json:"code"`
+	Message string    `json:"message"`
+}
+
+func NewError(code ErrorCode, msg string) *Error {
+	return &Error{
+		Code:    code,
+		Message: msg,
+	}
+}
+
+func (e *Error) Error() string {
+	return fmt.Sprintf("error: code = %v, message = %s", e.Code, e.Message)
+}
+
+func (e *Error) ToJSONString() string {
+	s, err := json.Marshal(e)
+	if err != nil {
+		return fmt.Sprintf(`{"code":0,"message":"Cannot marshal to json string for error: %s"}`, e.Error())
+	}
+	return string(s)
+}
+
+func WrapError(in error, format string, args ...interface{}) error {
+	if in == nil {
+		return nil
+	}
+
+	out, ok := in.(*Error)
+	if !ok {
+		out = NewError(ErrorCodeResultUnknown,
+			fmt.Sprintf("%s: %s", fmt.Sprintf(format, args...), in.Error()))
+	} else {
+		out.Message = fmt.Sprintf("%s: %s", fmt.Sprintf(format, args...), out.Message)
+	}
+	return out
+}
+
+func CombineErrors(errorList ...error) (retErr error) {
+	for _, err := range errorList {
+		if err != nil {
+			if retErr != nil {
+				retErr = fmt.Errorf("%v; %v", retErr, err)
+			} else {
+				retErr = err
+			}
+		}
+	}
+	return retErr
+}
+
+func GenerateFunctionErrorWithRollback(functionErr, rollbackErr error) error {
+	if functionErr != nil {
+		if rollbackErr != nil {
+			return NewError(ErrorCodeFunctionFailedRollbackFailed,
+				CombineErrors(functionErr, rollbackErr).Error())
+		}
+		return NewError(ErrorCodeFunctionFailedRollbackSucceeded,
+			functionErr.Error())
+	}
+	if rollbackErr != nil {
+		return NewError(ErrorCodeResultUnknown,
+			fmt.Sprintf("function succeeds but there is rollback error: %v", rollbackErr))
+	}
+
+	return nil
+}
+
+func UnmarshalGRPCError(err error) error {
```

Comment:
Can we just create an error with two errors? We can type convert grpc error back instead of depending on the error message format for parsing.

Reply:
If we use the customized error as the gRPC server output, the error will be wrapped as a gRPC error automatically and the gRPC client always receives gRPC errors. That's why I convert the error as JSON string then unmarshal it in gRPC client.

Reply:
It seems gRPC errors can contain more details. See https://grpc.io/docs/guides/error/ and https://jbrandhorst.com/post/grpc-errors/